File: app/Models/CachedData.php
<?php

namespace Irail\Models;

use Carbon\Carbon;

/**
 * Data which has been cached.
 *
 * @template T, the cached data type
 * @template-implements CachedData<T>
 */
class CachedData implements Cachable
{
    private string $key;
    private null|object|string|array|bool $value;
    private int $createdAt;
    private ?int $expiresAt;

    /**
     * @param string                        $key
     * @param object|array|string|bool|null $value
     * @param int                           $ttl
     */
    public function __construct(string $key, null|object|array|string|bool $value, int $ttl = 0)
    {
        $this->key = $key;
        $this->value = $value;
        $this->createdAt = time();
        if ($ttl > 0) {
            $this->expiresAt = time() + $ttl;
        } else {
            $this->expiresAt = null;
        }
    }

    public function setValue(mixed $value): void
    {
        $this->value = $value;
    }

    /**
     * The key of this cache entry.
     *
     * @return string
     */
    public function getKey(): string
    {
        return $this->key;
    }

    /**
     * The value of this cache entry.
     *
     * @return T
     */
    public function getValue(): mixed
    {
        return $this->value;
    }

    /**
     * @return Carbon the timestamp at which this entry was created
     */
    public function getCreatedAt(): Carbon
    {
        return Carbon::createFromTimestamp($this->createdAt);
    }

    /**
     * @return ?Carbon
     */
    public function getExpiresAt(): ?Carbon
    {
        return $this->expiresAt == null ? null : Carbon::createFromTimestamp($this->expiresAt);
    }

    /**
     * @param int $expiresAt
     * @return CachedData
     */
    public function setExpiresAt(int $expiresAt): CachedData
    {
        $this->expiresAt = $expiresAt;
        return $this;
    }


    /**
     * @return int get the age of this entry in seconds
     */
    public function getAge(): int
    {
        return time() - $this->createdAt;
    }

    public function getRemainingTtl(): int
    {
        return $this->expiresAt - time();
    }

    public function setTtl(int $ttl)
    {
        if ($ttl == 0) {
            $this->expiresAt = 0;
        } elseif ($ttl < 0) {
            $this->expiresAt = time();
        } else {
            $this->expiresAt = time() + $ttl;
        }
    }
}

File: src/main/java/be/irail/api/dto/CachedData.java
package be.irail.api.dto;

import java.time.Instant;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;

/**
 * Container for data that has been cached.
 * Stores the cached value along with its creation and expiration timestamps.
 *
 * @param <T> the type of the cached data
 */
public class CachedData<T> {
    private final T value;
    private final long createdAt;
    private long expiresAt;

    /**
     * Constructs a new CachedData entry.
     *
     * @param value the data to be cached
     * @param ttl the time-to-live in seconds
     */
    public CachedData(T value, int ttl) {
        this.value = value;
        this.createdAt = Instant.now().getEpochSecond();
        this.expiresAt = this.createdAt + ttl;
    }

    /**
     * Gets the cached value.
     * @return the cached data
     */
    public T getValue() {
        return this.value;
    }

    /**
     * Gets the timestamp when this entry was created.
     * @return the creation datetime
     */
    public OffsetDateTime getCreatedAt() {
        return OffsetDateTime.ofInstant(Instant.ofEpochSecond(this.createdAt), ZoneOffset.UTC);
    }

    /**
     * Gets the timestamp when this entry expires.
     * @return the expiration datetime
     */
    public OffsetDateTime getExpiresAt() {
        return OffsetDateTime.ofInstant(Instant.ofEpochSecond(this.expiresAt), ZoneOffset.UTC);
    }

    /**
     * Sets a new expiration timestamp for this entry.
     *
     * @param expiresAt the new expiration timestamp (epoch seconds)
     * @return this instance for chaining
     */
    public CachedData<T> setExpiresAt(long expiresAt) {
        this.expiresAt = expiresAt;
        return this;
    }

    /**
     * Calculates the current age of this entry in seconds.
     * @return the age in seconds
     */
    public long getAge() {
        return Instant.now().getEpochSecond() - this.createdAt;
    }
}
